metabase/metabase
# feat(sdk): override chart colors

Closes https://github.com/metabase/metabase/issues/42685

### Description

We want to give users options to override the chart colors they've pre-defined in their appearance settings using the `theme.colors.charts` SDK theme settings.

The `charts` array is an array of chart colors, which can either be a color string (most commonly used) or an object that defines a base color with an optional `shade` and `tint` variations. Note that when the shades and tints are not explicitly defined, they're automatically generated by the existing logic in `lib/colors/palette.ts` as part of the `aliases` object.

```ts
{
  colors: {
    charts: [
      "#e74c3c",
      "#6ab04c",
      { base: "#4834d4" },
      { base: "#be2edd", lighter: "#e056fd" },
      { base: "#30336b", darker: "#130f40" },
    ],
  }
}
```

### Limitations

**We cannot override the color of chart series yet.** The database stores the hex colors in `series_settings.[key].color` rather than a stable identifier that we can reliably correlate or map to the chart colors. We store the hex colors at that point in time in the database rather than a key to lookup the colors, so we won't be able to globally override all chart colors for now.

Currently, this PR works for charts **without series color overrides applied**. It should work for all viz, including cartesian charts (bar, line, area), pie chart, funnel, waterfall and row chart. For cartesian charts, the default preferred colors will be applied according to summarization: count = `accent0`, sum = `accent1`, average = `accent2`. The above limitation will be addressed later.

<img src="https://github.com/metabase/metabase/assets/4714175/6417dd0a-44f7-492a-b750-1202af3af5e6" width="300">

### How to verify

- Go to the `test-sdk-chart-colors` branch in the demo app
- Go to `AppProvider.tsx`, and update `theme.colors.charts` to the colors you want to test. There can only be a maximum of 8 accent colors.
- Go to the Kitchen Sink. The colors of the charts should change according to the chart colors.

### Demo

<img src="https://github.com/metabase/metabase/assets/4714175/6b8bef47-5e8b-490a-8896-a8eb8e303ffb" width="300">

### Checklist
